Add check for indent size violations

igelbox opened the pull request:
#### PR checklist

- [x] Addresses an existing issue: #2814
- [x] New feature, bugfix, or enhancement
  - [x] Includes tests
- [ ] Documentation update

#### Overview of change:
Fixed the #2814 issue.
Updated logic to make it smarter, so now it's able to check/correct indentation even after fixing wrong characters. TSlint 5.7.0 may require a second pass to fix indentation in these cases.

#### Is there anything you'd like reviewers to focus on?
Changed test files, e.g. test/rules/indent/spaces-2/test.ts.fix

palantirtech commented:
Thanks for your interest in [palantir/tslint](https://github.com/palantir/tslint), @igelbox! Before we can accept your pull request, you need to sign our contributor license agreement - just visit https://cla.palantir.com/ and follow the instructions. Once you sign, I'll automatically update this pull request.

ajafff commented on test/rules/indent/spaces-2/test.ts.lint:
```diff
@@ -159,4 +159,9 @@ var arr2 = [
+var arr3 = [
+    1
+~~~~ [0]
+];
```
this needs another test for nesting:
```ts
var arr3 = [
    {
        foo,
    }
]
```

Also add a test with more than 4 spaces indent:
```ts
var foo =
          bar;
```

igelbox replied:
As for the "another" `var arr3 = [` case, this(fixed) implementation cannot fix it correctly due to, there isn't some kind of multiline-ready fix. So, after the `--fix` this example will look like:
```js
var arr3 = [
  {
    foo,
    }
]
```
The latest `}` won't be treated by the tslint as an issue due to, it conforms naive rules of this(fixed) implementation.
To fix that issue correctly, we should dig into the AST of a linting code fragment much deeper. Otherwise, a naive implementation won't feel the difference between your example and a single-line code-style error, e.g.:
```js
var arr3 = [
    { // the error is only here
    foo,
  }
]
```
which may be fixed by a naive multiline-ready fixer as:
```js
var arr3 = [
  { // the error is only here
  foo,
}
]
```

So, seems like we'd better not fix indent size violations automatically. Instead of it, we should only report a notification and rely on only a manual fixing in these cases.

FeaRCODE commented:
Do we have an estimate of when this might be merged? I work with: `"indent": [ true, "tabs", 4 ]` and it's annoying not to be able to autofix indention issues created by angular cli. Thanks :)

giladgray commented:
@igelbox please update this branch so we can review it, or close it if not relevant.

igelbox commented:
@giladgray Maybe I'm missing something, but, I cannot see any non-addressed review point nor any merge conflict.
So, could you please tall me what kind of action should I do to update the branch?

johnwiseheart commented:
@igelbox You just need to push a small (or empty) commit to force the tests to run again. Thanks!

adidahiya commented:
I think this needs a lot more tests; autofixing indentation is nontrivial (it's essentially the bulk of the functionality provided by a code formatter), so merging this should not be taken lightly. See the lengthy discussion in https://github.com/palantir/tslint/pull/3395

don-smith commented:
Unfortunately #2814 aims to address 2 different issues: detecting the violation and fixing it. Because many of us don't need tslint to `--fix` the violation (yet), the larger formatting effort should be managed separately so it doesn't block smaller bug fixes. If this PR successfully detects the violation, it should be merged.

igelbox commented:
That's weird. When I ran `yarn test` on the branch locally, tests were successful. However, when I switched to the master, ran `yarn`, switched back, and ran `yarn` once again - tests are failing just line in CI.
Well, the branch looks too outdated, so, lemme merge master into it.

NicoAiko commented:
What's going on here? This has been open for almost 2 months...
